"Using Signals in Fluid" Recipe for FF.com (#11161)
microsoft/FluidFramework: docs/content/docs/recipes/signals.md

To track the client's mouse position, we use the `mousemove` event to obtain the client's x and y coordinates. To alert the connected clients of a mouse position change, we then submit a `mouseSignalType` signal with the client's userID and updated position value as the payload:

```typescript
window.addEventListener("mousemove", (e) => {
    const position: IMousePosition = {
        x: e.clientX,
        y: e.clientY,
    };
    this.sendMouseSignal(position);
});
```
```typescript
/**
* Alert all connected clients that there has been a change to a client's mouse position
*/
private sendMouseSignal(position: IMousePosition) {
    this.signaler.submitSignal(
        MouseTracker.mouseSignalType,
        { userId: this.audience.getMyself()?.userId, pos: position },
    );
}
```
But what is the point of sending a signal if nobody is listening to it? To listen to the `mouseSignalType`, we use `Signaler`'s `onSignal` function. We then update the local data using the payload information from the signal. To display the new presence data, we then emit `mousePositionChanged` to let the view know to re-render:

```typescript

> this.signaler.onSignal(MouseTracker.mouseSignalType, (clientId: string, local: boolean, payload: Jsonable) => {

Reviewer: Also, for what it is worth, the example app that was linked to in the first paragraph of the article doesn't have these type specifications in the MouseTracker.ts file. For example, this line is like this in the example app: 

```
this.signaler.onSignal(MouseTracker.mouseSignalType, (clientId, local, payload) => {
```

> this.signaler.onSignal(MouseTracker.mouseSignalType, (clientId: string, local: boolean, payload: Jsonable) => {

Author: Good point, my thoughts on the whole `Jsonable` usage is still something I'm working through. [Here](https://github.com/microsoft/FluidFramework/pull/11161#discussion_r929095282) is a link to the discussion I'm having with Matt about the issue, feel free to leave thoughts :)

  this.onMouseSignalFn(clientId, payload);
});
```
```typescript
private readonly onMouseSignalFn = (clientId: string, payload: Jsonable) => {
    const userId: string = payload.userId;
    const position: IMousePosition = payload.pos;

    let clientIdMap = this.posMap.get(userId);
    if (clientIdMap === undefined) {
        clientIdMap = new Map<string, IMousePosition>();
        this.posMap.set(userId, clientIdMap);
    }
    clientIdMap.set(clientId, position);
    this.emit("mousePositionChanged");
};
```

> ## FocusTracker

Reviewer: This section should start with a brief explanation of exactly what kind of thing has (or doesn't have) focus. Is it an HTML page? A `<div>`? What?

Let's now look at how the `FocusTracker` class uses the `Signaler` DataObject to share user focus infomation.

The class defines the `focusSignalType` that will be sent and listened to the connected clients when there is a presence change:

```typescript
export class FocusTracker extends TypedEventEmitter<IFocusTrackerEvents> {
    private static readonly focusSignalType = "changedFocus";
    /*...*/
}
```

> The class then initializes a local map of boolean values for all of the connected clients. The boolean denotes whether or not the client has focus or not:

Reviewer: What does "the client has focus" mean? "Focus" is a UI term, but "client" isn't.

    if (focusClientIdMap !== undefined) {
        focusClientIdMap.delete(clientId);
        if (focusClientIdMap.size === 0) {
            this.focusMap.delete(member.userId);
        }
    }
    this.emit("focusChanged");
});
```
To track the local client's focus status, we use the `focus` and `blur` events to know when the focus boolean must be updated. To alert the connected clients of this focus change, we then submit a `focusSignalType` signal with the client's userID and updated focus status as the paylod:

```typescript
window.addEventListener("focus", () => {
    this.sendFocusSignal(true);
});
window.addEventListener("blur", () => {
    this.sendFocusSignal(false);
});
```
```typescript
private sendFocusSignal(hasFocus: boolean) {
    this.signaler.submitSignal(
        FocusTracker.focusSignalType,
        { userId: this.audience.getMyself()?.userId, focus: hasFocus },
    );
}
```
To make sure all connected clients are notified of this focus change, we use the `onSignal` function to listen to the `focusSignalType`. We then update the local data using the payload information from the signal and emit a `focusChanged` event to re-render:

```typescript

> this.signaler.onSignal(FocusTracker.focusSignalType, (clientId, local, payload) => {

Reviewer: The corresponding line in the MouseTracker above specified data types:

```
(clientId: string, local: boolean, payload: Jsonable) 
```

Is there a reason for the difference? I think specifying the types makes the code more readable/understandable, although the actual example app that is linked to at the top of this article doesn't specify the types. (Maybe it should?)

```typescript
private static readonly focusRequestType = "focusRequest";
```
`FocusTracker` then sends this signal request immediately after the new client connects to the container. Once again, the `onSignal` function is used to listen to this `focusRequestType` signal and each client responds to the signal request with their current focus status:

```typescript
this.signaler.onSignal(FocusTracker.focusRequestType, () => {
    this.sendFocusSignal(document.hasFocus());
});

container.on("connected", () => {
    this.signaler.submitSignal(FocusTracker.focusRequestType);
});
```

## View

In `app.ts`, we create instances of the `MouseTracker` and the `FocusTracker` to use in our application:

```typescript
/*...*/

async function start(): Promise<void> {
    /*...*/

    // Render presence information for audience members
    const contentDiv = document.getElementById("focus-content") as HTMLDivElement;
    const mouseContentDiv = document.getElementById("mouse-position") as HTMLDivElement;
    const focusTracker = new FocusTracker(
        container,
        services.audience,

>         container.initialObjects.signaler as Signaler,

Reviewer: When the initialObjects was defined, the type was specified as "Signaler", so why do we need to cast it to Signaler?

---
title: Using Signals in Fluid
menuPosition: 5
draft: true
---

> In this recipe, we will go over the `PresenceTracker` example to learn how the `Signaler` DataObject is used in a Fluid application to deal with a couple user presence scenarios. The two forms of presence covered in this example are focus tracking and mouse tracking. This article will go over how the `Signaler` DataObject is used in both the `MouseTracker` and `FocusTracker` classes to achieve user presence in a Fluid application. You can find the completed code for this example in the Fluid Repo [here](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker).

Reviewer 2: ```suggestion
In this recipe, we will go over the `PresenceTracker` example to learn how the `Signaler` DataObject is used in a Fluid application to share user presence information between collaborators. We'll cover how the `Signaler` DataObject is used in both the `MouseTracker` and `FocusTracker` classes to share mouse position and focus state. You can find the completed code for this example in the Fluid Repo [here](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker).
```

## Creation

The `Signaler` DataObject is first included in the `initialObjects` of the Fluid `ContainerSchema` which defines which shared objects will be at our disposal in the new container.

```typescript
const containerSchema: ContainerSchema = {
    initialObjects: {
        /* [id]: DataObject */
        signaler: Signaler,
    },
};
```

The `FluidContainer` is then created (or fetched if it already exists) alongside the `services` object, since audience will be required in this application. To learn more about how to create a new Fluid container and load an existing container, see our [DiceRoller Tutorial](https://fluidframework.com/docs/start/tutorial/).

## MouseTracker

> Let's look at how the `MouseTracker` class uses the `Signaler` DataObject to achieve the first form of user presence within the application.

Reviewer 2: ```suggestion
Let's look at how the `MouseTracker` class uses the `Signaler` DataObject to share mouse position information.
```

The class defines the `mouseSignalType` that will be sent and listened to the connected clients when there is a presence change:

```typescript
export interface IMouseTrackerEvents extends IEvent {
    (event: "mousePositionChanged", listener: () => void): void;

> }

Reviewer 2: You don't talk about this event until later -- maybe omit from the code snippets here until you're ready to talk about it below.

export class MouseTracker extends TypedEventEmitter<IMouseTrackerEvents> {
    private static readonly mouseSignalType = "positionChanged";

    /*...*/
}
```

> The class also initializes a local map of `IMousePosition` values for all of the connected clients. Each userID can have multiple clientIDs (e.g. same user on separate devices), which explains the nested `Map`. This local map is what populates the view and what will be updated on `mouseSignalType` signals:

Reviewer 2: Maybe reference the Audience article to explain the user->client nesting:

https://fluidframework.com/docs/build/audience/

> The class also initializes a local map of `IMousePosition` values for all of the connected clients. Each userID can have multiple clientIDs (e.g. same user on separate devices), which explains the nested `Map`. This local map is what populates the view and what will be updated on `mouseSignalType` signals:

Reviewer 2: ```suggestion
The class also initializes a local map (`posMap`) of `IMousePosition` values for all of the connected clients. Each userID can have multiple clientIDs (e.g. same user on separate devices), which explains the nested `Map`. This local map is what populates the view and what will be updated on `mouseSignalType` signals:
```

```typescript
export interface IMouseTrackerEvents extends IEvent {
    (event: "mousePositionChanged", listener: () => void): void;
}

export interface IMousePosition {
    x: number;
    y: number;
}

export class MouseTracker extends TypedEventEmitter<IMouseTrackerEvents> {
    private static readonly mouseSignalType = "positionChanged";

    /**
     * Local map of mouse position status for clients
     *
     * ```
     * Map<userId, Map<clientid, position>>
     * ```
     */
    private readonly posMap = new Map<string, Map<string, IMousePosition>>();

    /*...*/
}
```

> Now that we have clarity about what `posMap` and `mouseSignalType` are, we can move to the constructor to see what other behavior our `MouseTracker` has.

Reviewer 2: I like that you lead with "here's the core data that the tracker is holding (posMap)".  To segue from the data to the constructor requirements, maybe something like

```suggestion
Now that we have clarity about what `posMap` and `mouseSignalType` are, we need to use the `audience` and `Signaler` to send, receive, and interpret the positions.  We provide these in the constructor.
```

We'll begin by looking at the constructor arguments. `MouseTracker` takes in the `audience` and the `Signaler` instance from `initialObjects`:

```typescript
export interface IMouseTrackerEvents extends IEvent {
    (event: "mousePositionChanged", listener: () => void): void;
}

export class MouseTracker extends TypedEventEmitter<IMouseTrackerEvents> {
    /*...*/

    public constructor(

>       public readonly audience: IServiceAudience<IMember>,

Reviewer 2: This really shouldn't be public I think, but maybe something to worry about later.

      private readonly signaler: Signaler,
    ){
      super();

      /*...*/
    }

    /*...*/
}
```

> Whenever a member leaves the audience (e.g., a client disconnects from the container), we would want to remove their presence from our local data. After this is done, we would have to let the view know that the local presence data has changed. To do this, we emit our `mousePositionChanged` event so the view knows to re-render:

Reviewer 2: Prior to talking about members leaving the audience, maybe talk about the `mousePositionChanged` event more generally -- e.g. "Whenever there is any change to the positions we're tracking, we'll fire a `mousePositionChanged` event to notify listeners that updated data is available."

```typescript
export interface IMouseTrackerEvents extends IEvent {
    (event: "mousePositionChanged", listener: () => void): void;
}

export class MouseTracker extends TypedEventEmitter<IMouseTrackerEvents> {
    /*...*/

    public constructor(
      public readonly audience: IServiceAudience<IMember>,
      private readonly signaler: Signaler,
    ){
      super();

>       this.audience.on("memberRemoved", (clientId: string, member: IMember) => {

Reviewer 2: In general, probably scope code snippets to just the part you're actively talking about in that section to help the reader focus on it specifically.

        const clientIdMap = this.posMap.get(member.userId);
        if (clientIdMap !== undefined) {
            clientIdMap.delete(clientId);

            //If UserID has no connected clients, remove user from local data
            if (clientIdMap.size === 0) {
                this.posMap.delete(member.userId);
            }
        }
        this.emit("mousePositionChanged");
      });

        /*...*/
    }

    /*...*/
}
```

> We know that `MouseTracker` must be able to share information to all connected clients about where the current client's mouse position is. This is where `Signaler` comes in handy, as it specializes in communicating transient data to connected clients within a Fluid application.

Reviewer 2: ```suggestion
`MouseTracker` must share information to all connected clients about where the local client's mouse position is. This is where `Signaler` comes in handy, as it specializes in communicating transient data to connected clients within a Fluid application.
```

> To track the client's mouse position, we use the `mousemouve` event to obtain the client's x and y coordinates. To alert the connected clients of a mouse position change, we then submit a `mouseSignalType` signal with the client's userID and updated position value as the paylod:

Reviewer 2: ```suggestion
To track the client's mouse position, we use the `mousemove` event to obtain the client's x and y coordinates. To alert the connected clients of a mouse position change, we then submit a `mouseSignalType` signal with the client's userID and updated position value as the payload:
```

      window.addEventListener("mousemove", (e) => {
        const position: IMousePosition = {
            x: e.clientX,
            y: e.clientY,
        };
        this.sendMouseSignal(position);
      });

        /*...*/
    }

    /**
     * Alert all connected clients that there has been a change to a client's mouse position
     */
    private sendMouseSignal(position: IMousePosition) {
        this.signaler.submitSignal(
            MouseTracker.mouseSignalType,
            { userId: this.audience.getMyself()?.userId, pos: position },
        );
    }

    /*...*/
}
```
But what is the point of sending a signal if nobody is listening to it? To listen to the `mouseSignalType`, we use `Signaler`'s `onSignal` function. We then update the local data using the payload information from the signal. To display the new presence data, we then emit `mousePositionChanged` to let the view know to re-render:

```typescript
export class MouseTracker extends TypedEventEmitter<IMouseTrackerEvents> {
    /*...*/

    public constructor(
        public readonly audience: IServiceAudience<IMember>,
        private readonly signaler: Signaler,
    ){
        super();

        /*...*/

>         this.signaler.onSignal(MouseTracker.mouseSignalType, (clientId, local, payload) => {

Reviewer 2: Maybe put types on the callback parameters to help folks understand what these are exactly.

          this.onMouseSignalFn(clientId, payload);
        });

        window.addEventListener("mousemove", (e) => {
            const position: IMousePosition = {
                x: e.clientX,
                y: e.clientY,
            };
            this.sendMouseSignal(position);
        });

        /*...*/
    }

    /*...*/

>     private readonly onMouseSignalFn = (clientId: string, payload: any) => {

Reviewer 2: Similar here - can we put a more-specific type than `any`?  Even if we have to cast to it, it will help readers understand what we expect the payload to contain.

Author: I'm sort of stumped on what type should be used here in place of `any`. The payload is previously defined in SignalListener as having type `Jsonable` so that's what I currently switched it to. But, the more I read into `Jsonable` type [here](https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/datastore-definitions/src/jsonable.ts), I see that it says 

>Using `Jsonable` (with no type parameters) or `Jsonable<any>` is just a type alias for `any` and should not be used if type safety is desired.

A payload is just any serializable value, would there be a more appropriate type to use rather than `Jsonable`? It would also take more explaining to the customer what the type means since `Jsonable` is defined in @fluidframework/datastore-definitions and is probably foreign to the average consumer.

Author: Make sense, I will do the same for the FocusTracker payload as well. Thanks!

To achieve this pattern, `FocusTracker` defines the `focusRequestType` that will be sent to request the focus status of all the connected clients:

```typescript
export class FocusTracker extends TypedEventEmitter<IFocusTrackerEvents> {
    private static readonly focusRequestType = "focusRequest";

    /*...*/
}
```
`FocusTracker` then sends this signal request immediately after the new client connects to the container. Once again, the `onSignal` function is used to listen to this `focusRequestType` signal and each client responds to the signal request with their current focus status:

```typescript
export class FocusTracker extends TypedEventEmitter<IFocusTrackerEvents> {
    private static readonly focusRequestType = "focusRequest";

    /*...*/

    public constructor(
        container: IFluidContainer,
        public readonly audience: IServiceAudience<IMember>,
        private readonly signaler: Signaler,
    ) {
        super();

        /*...*/

        this.signaler.onSignal(FocusTracker.focusRequestType, () => {
            this.sendFocusSignal(document.hasFocus());
        });

>         container.on("connected", () => {

Author: Will keep in mind

            this.signaler.submitSignal(FocusTracker.focusRequestType);
        });
    }
}
```

## View

In `app.ts`, we create instances of the `MouseTracker` and the `FocusTracker` to use in our application:

```typescript
/*...*/

async function start(): Promise<void> {
    /*...*/

    // Render presence information for audience members
    const contentDiv = document.getElementById("focus-content") as HTMLDivElement;
    const mouseContentDiv = document.getElementById("mouse-position") as HTMLDivElement;
    const focusTracker = new FocusTracker(
        container,
        services.audience,
        container.initialObjects.signaler as Signaler,
    );
    const mouseTracker = new MouseTracker(
        services.audience,
        container.initialObjects.signaler as Signaler,
    );

    /*...*/
}

start().catch(console.error);
```

> The view then renders the focus data by using `renderFocusPresence`, which uses the local focus status map to display which users are in focus and which users are not. It also renders the mouse position data by using `renderMousePresence`, which uses the local position map to display the name of each user where they currently are in the window:

Reviewer 2: Maybe show the snippets for the render functions, or a portion of them?  In particular would probably be good to show registering for the events to re-render.
